library/README.md

```diff
@@ -1,6 +1,66 @@
+| ipfs::ctx::JsonParser             | Parses JSON (String -> ipfs::DagJsonValue)                           | ipfs::ctx::NlohmannJsonParser                           | Uses N. Lohmann's [JSON for Modern C++](https://github.com/nlohmann/json)                                               |
+| ipfs::ctx::CborParser             | Parses CBOR (Bytes -> ipfs::DagCborValue)                            | ipfs::ctx::NlohmannCborParser                           | Uses N. Lohmann's [JSON for Modern C++](https://github.com/nlohmann/json), but for CBOR                                 |
+| ipfs::gw::Requestor               | A chain-of-responsibility for processing IPFS requests               | ipfs::gw::default_requestor() returns common requestors | If you don't provide an argument, this chain does no caching.                                                           |  
+| MimeTypeDeduction                 | A function that guesses the mime type of a doc given its head & path | ipfs::util::TrivialMimeGuess                            | Doesn't even look at file content, just uses extension. Fine for cases that don't really need the mime type.            |
+| UrlUnescaping                     | Given a URL component (i.e. in-between `/`s), deal with % escaping   | ipfs::util::RoughlyUnescapeUrlComponent                 | Not a complete implementation, but deals with some common cases.                                                        |
+| ipfs::crypto::SignatureVerifier   | Verifies a public key signature for a particular type of key         | ipfs::crypto::OpensslSignatureVerifier                  | Uses OpenSSL, but only for RSA & Ed25519 keys (others are dropped)                                                      |
+
+### Get a partition
+
+`ipfs::Client::partition()` takes an arbitrary string to use as a key for that partition. 
+The scheme for the keys is whatever you'd like it to be, but if you pass in an identical key you'll get the same partition back each time.
+If it's the first time for that key, a new object is created, including an isolated view of the DAGs.
+
+If you are not in a security-critical case, it's entirely reasonable to pass in `""` every time.
+You might also consider using URL origins to define partitioning.
+
+### Make a request
+
+`std::make_shared<IpfsRequest>(path, callback)`
+
+* path - A string of the form /ipfs/<cid>/path/to/file or /ipns/<key or hostname>/path/to/file
+* callback - What to call when a final response is formed.
+
+Pass the request to `partition->build_response()`
+
+### Drive your HttpApi and DnsTxtLookup implementations
+
+Many implementations will need some sort of event loop. Do that.
+
+### Handle response
+
+The callback you passed to create the request. 
+Be sure to check the second parameter's (the response's) status_ field, which contains HTTP statuses like:
+
+* 200 - success
+* 404 - unable to find it from known gateways (possibly timed out)
```

Comment:
404 should not be used for timeouts, but for content that is provably not in a dag, maybe use  https://specs.ipfs.tech/http-gateways/path-gateway/#504-gateway-timeout here?

Reply:
That makes sense. I'll have to be careful because I already use 504 to indicate a gateway timed out (rather than every gateway I know about failed in one way or another).

Reply:
It kinda sounds like I might want to use 502 for this 

> For example, when gateway failed to find any providers for requested data.